utils/gridlink_impl.c.src

```diff
@@ -179,113 +142,158 @@ cellarray_DOUBLE * gridlink_DOUBLE(const int64_t NPART,
-                    orig_index = my_realloc(lattice[icell].original_index, sizeof(lattice[icell].original_index[0]), expected_n, "lattice.x");
-
-                    lattice[icell].x = (posx == NULL) ? lattice[icell].x:posx;
-                    lattice[icell].y = (posy == NULL) ? lattice[icell].y:posy;
-                    lattice[icell].z = (posz == NULL) ? lattice[icell].z:posz;
-                    lattice[icell].original_index = (orig_index == NULL) ? lattice[icell].original_index:orig_index;
-
-                    w_alloc_status = EXIT_SUCCESS;
-                    for(int w = 0; w < lattice[icell].weights.num_weights; w++){
-                        DOUBLE *newweights = (DOUBLE *) my_realloc(lattice[icell].weights.weights[w], sizeof(DOUBLE), expected_n, "lattice.weights");
-                        if(newweights == NULL){
-                            w_alloc_status = EXIT_FAILURE;
-                        } else {
-                            lattice[icell].weights.weights[w] = newweights;
-                        }
-                    }
+            // Now do the same for the weights
+            for(int w = 0; w < num_weights; w++){
+                lattice[icell].weights.weights[w] = ((DOUBLE *) WEIGHTS->weights[w]) + cellstarts[icell];
+            }
+        }
+    }//end of loop over totncells
+    
+    free(cellstarts);
 
-                    if(posx == NULL || posy == NULL || posz == NULL || orig_index == NULL ||w_alloc_status == EXIT_FAILURE) {
-                        expected_n--;
-                    }
-                } while(expected_n > nallocated[icell] && (posx == NULL ||
-                                                           posy == NULL ||
-                                                           posz == NULL ||
-                                                           orig_index == NULL ||
-                                                           w_alloc_status == EXIT_FAILURE));
-
-                if(expected_n == nallocated[icell]) {
-                    /*realloc failed. free memory and return */
-                    fprintf(stderr,"In %s> Reallocation failed,  randomly subsampling the input particle set (currently at %"PRId64" particles) might help\n",
-                            __FUNCTION__, NPART);
-                    free_cellarray_DOUBLE(lattice, totncells);
-                    free(nallocated);
-                    return NULL;
+    // Each thread is going to loop over all particles, but only process those in its cell domain
+#if defined(_OPENMP)
+    #pragma omp parallel
+#endif
+    {
+        #if defined(_OPENMP)
+        int tid = omp_get_thread_num();
+        #else
+        int tid = 0;
+        #endif
+        int64_t cstart = totncells*tid/Nthread;
+        int64_t cend = totncells*(tid+1)/Nthread;
+        // Second loop over particles: having computed the starts of each cell, fill the cells
```

Comment:
Probably needs a #pragma omp for annotation

Reply:
This is intentional---all threads loop over all particles, but only process ones that fall in their cell domain.  There's a comment about this above; I will move it closer.